rfcs/0012-compaction-state-persistence.md

```diff
@@ -0,0 +1,665 @@
+
+```
+/000000001.compactor  # First compactor state
+/000000002.compactor  # Updated state after compactions
+/000000003.compactor  # Current state
+```
+
+#### **CompactionState Structure**
+The persistent state contains the complete view of all compaction activity:
+
+```rust
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CompactionState {
+    /// Fencing token to ensure single active compactor
+    /// Incremented each time a new compactor takes control
+    pub compactor_epoch: u64,
+    
+    /// Compactor state identifier. This would be used for creating
+    /// compactor files and CAS updates
+    pub compactor_state_id: u64,
+    
+    /// All currently active compactions indexed by ID
+    /// Includes queued, running, and recently completed compactions
+    pub active_compactions: BTreeMap<CompactionId, Compaction>,
+        
+    /// Timestamp when this state was created/last updated
+    pub state_timestamp: DateTime<Utc>,
+}
+```
+The section below is under discussion here: https://github.com/slatedb/slatedb/pull/695/files#r2239561471
+
+### Protocol for State Management of Manifest and CompactionState
+
+This a proposol for Statement Management of Manifest and CompactionState. The protocol is based on the following principals:
+
+- Compaction is an entity owned by the Compactor. Therefore, the `compactor_epoch` and the compacted `SortedRuns` are also owned by the Compactor.
+- CAS based update of the .compactor file ensures consistent view of the compactionState to all the compactor processes.
+
+#### On startup...
+1. Compactor fetches the latest .manifest file (00005.manifest).
+
+2. Compactor now fetches the latest .compactor file (00005.compactor). 
+
+3. It builds a dirty compactionState by merging L0 SSTs and SortedRuns across both files.
+
+    - Copy SortedRuns from .compactor file to compactionState
+
+    - Add and delete L0 SSTs of .manifest in compactionState
+
+4. Compactor increments `compactor_epoch` in the dirty CompactionState and writes the dirty CompactionState to the next sequential .compactor position.(00006.compactor). Two level validation:
+
+    Epoch Check fails, compactor is fenced
+
+    File version check (in-memory and remote object store), If 00006.compactor exists, 
+
+      - If latest .compactor compactor_epoch > current compactor's epoch, die (fenced)
+
+      - If latest .compactor compactor_epoch == current compactor's epoch, die (fenced)
+
+      ( in case of external processes like CLI sending a compaction request, the compaction persistence in the CompactionState would be done by the active Compactor)
+
+      - If latest .compactor compactor_epoch < current compactor's epoch, increment the .compactor file ID by 1 and retry. This process would continue until successful compactor write.
+      ( The current active compactor Job would have updated the .compactor file)
+
+
```

Comment:
It needs to write the manifest after writing the compactor file. That's mentioned in the summary below, but not here